### PR TITLE
Coot S01E02

### PR DESCRIPTION
**Pull request by shrit**
This could be merged fast.

`arma::repmat -> repmat`
`arma::randu  -> randu`

`repmat` function name only exist in Matlab, so no harm in removing arma,
Opencv has the same `randu` name so it could be tricky, especially if someone is doing object detection, unless if they use `cv::randu` in their code.

### CODE REVIEW

**Comment by shrit** [comment]
I thought for `randu` it would work, but it seems that the template parameter would not be considered in this context, only function parameters. 

This should pass for `repmat` in this case

**Comment by shrit** [comment]
maybe someone needs to propose this to the C++ committee, to change ADL to ATDL :stuck_out_tongue: 
We need to think about functions that do not take `MatType` as a parameter.
Hopefully we can merge this before Monday

File: src/mlpack/methods/ann/layer/add_impl.hpp [hunk 1]
```diff
@@ -71,7 +71,7 @@ AddType<MatType>::operator=(AddType&& other)
 template<typename MatType>
 void AddType<MatType>::Forward(const MatType& input, MatType& output)
 {
-  output = input + arma::repmat(arma::vectorise(weights), 1, input.n_cols);
+  output = input + repmat(arma::vectorise(weights), 1, input.n_cols);
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/add_impl.hpp:74]: You could change `arma::vectorise` to `vectorise` too, or in another PR, up to you. :+1:

File: src/mlpack/methods/ann/augmented/tasks/copy_impl.hpp [hunk 2]
```diff
@@ -69,7 +69,7 @@ inline void CopyTask::Generate(arma::field<arma::mat>& input,
     arma::colvec vecInput = arma::randi<arma::colvec>(
       size, arma::distr_param(0, 1));
     arma::colvec vecLabel = arma::conv_to<arma::colvec>::from(
-      arma::repmat(vecInput, nRepeats, 1));
+      repmat(vecInput, nRepeats, 1));
```
Comment (rcurtin) [inline src/mlpack/methods/ann/augmented/tasks/copy_impl.hpp:72]:
```suggestion
        repmat(vecInput, nRepeats, 1));
```
Looks like there was a style "error" here and two lines above, feel free to fix it. :+1:

File: src/mlpack/methods/ann/layer/batch_norm_impl.hpp [hunk 3]
```diff
@@ -216,26 +216,26 @@ void BatchNormType<MatType>::Forward(
     // Calculate mean and variance over all channels.
     MatType mean = arma::sum(arma::sum(inputTemp, 2), 0) / m;
     variance = arma::sum(arma::sum(arma::pow(
-        inputTemp.each_slice() - arma::repmat(mean,
+        inputTemp.each_slice() - repmat(mean,
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/batch_norm_impl.hpp:219]: You can probably bring `repmat(mean, inputSize, 1)` all on one line now---probably easier to read too.

File: src/mlpack/methods/ann/layer/batch_norm_impl.hpp [hunk 4]
```diff
@@ -260,13 +260,13 @@ void BatchNormType<MatType>::Forward(
         const_cast<MatType&>(output).memptr(), inputSize, size,
         batchSize * higherDimension, false, false);
 
-    outputTemp.each_slice() -= arma::repmat(runningMean.t(),
+    outputTemp.each_slice() -= repmat(runningMean.t(),
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/batch_norm_impl.hpp:263]: Probably a lot of these lines can be shortened to one line now too.

File: src/mlpack/methods/lsh/lsh_search_impl.hpp [hunk 5]
```diff
@@ -270,7 +270,7 @@ void LSHSearch<SortPolicy, MatType>::Train(MatType referenceSet,
     // and the corresponding offset be 'offset_i'.  Then the key of a single
     // point is obtained as:
     // key = { floor((<proj_i, point> + offset_i) / 'hashWidth') forall i }
-    arma::mat offsetMat = arma::repmat(offsets.unsafe_col(i), 1,
+    arma::mat offsetMat = repmat(offsets.unsafe_col(i), 1,
                                        this->referenceSet.n_cols);
```
Comment (rcurtin) [inline src/mlpack/methods/lsh/lsh_search_impl.hpp:274]: This could probably all fit on one line now.

**Review by rcurtin — COMMENTED** [review]
The changes look good to me (let's see if the tests pass, I think they should).  The main comment I have is tedious: lots of the lines can be reflowed for easier readability now, since they are shorter.

**Comment by rcurtin** [comment]
I suppose maybe we will need to make a `SafeRandu()` function or similar, that dispatches to Armadillo or Bandicoot accordingly.  Perhaps sometime a long time from now we can use C++20: https://stackoverflow.com/questions/2953684/why-doesnt-adl-find-function-templates (apparently, ADL for functions with explicit template arguments works; I didn't check to see if it works right).

File: src/mlpack/methods/ann/layer/add_impl.hpp (hunk 1 above)
Comment (shrit) [reply src/mlpack/methods/ann/layer/add_impl.hpp:74]: in another PR, will make them simple short and fast

File: src/mlpack/methods/lsh/lsh_search_impl.hpp (hunk 5 above)
Comment (shrit) [reply src/mlpack/methods/lsh/lsh_search_impl.hpp:274]: unfortunately not for this one, it will be 87 chars :worried:

**Comment by rcurtin** [comment]
@mlpack-jenkins test this please

File: src/mlpack/methods/ann/layer/batch_norm_impl.hpp [hunk 6]
```diff
@@ -216,26 +216,24 @@ void BatchNormType<MatType>::Forward(
...
-        inputSize, 1) + eps);
+    outputTemp.each_slice() /= arma::sqrt(repmat(variance, inputSize, 1) + eps);
 
     // Re-used in backward propagation.
     normalized.set_size(arma::size(inputTemp));
     normalized = outputTemp;
 
-    outputTemp.each_slice() %= arma::repmat(gamma.t(),
+    outputTemp.each_slice() %= repmat(gamma.t(),
         inputSize, 1);
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/batch_norm_impl.hpp:235]:
```suggestion
    outputTemp.each_slice() %= repmat(gamma.t(), inputSize, 1);
```

File: src/mlpack/methods/ann/layer/batch_norm_impl.hpp [hunk 7]
```diff
@@ -216,26 +216,24 @@ void BatchNormType<MatType>::Forward(
...
-    outputTemp.each_slice() += arma::repmat(beta.t(),
+    outputTemp.each_slice() += repmat(beta.t(),
         inputSize, 1);
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/batch_norm_impl.hpp:237]:
```suggestion
    outputTemp.each_slice() += repmat(beta.t(), inputSize, 1);
```

File: src/mlpack/methods/linear_svm/linear_svm_function_impl.hpp [hunk 8]
```diff
@@ -171,7 +171,7 @@ LinearSVMFunction<MatType, ParametersType>::Evaluate(
     // On calculating the score, we add `b_i` term to each element of
     // `i_th` row of `scores`.
     scores = parameters.rows(0, dataset.n_rows - 1).t() * dataset
-        + arma::repmat(parameters.row(dataset.n_rows).t(), 1,
+        + repmat(parameters.row(dataset.n_rows).t(), 1,
         dataset.n_cols);
```
Comment (rcurtin) [inline src/mlpack/methods/linear_svm/linear_svm_function_impl.hpp:175]:
```suggestion
        + repmat(parameters.row(dataset.n_rows).t(), 1, dataset.n_cols);
```

File: src/mlpack/methods/linear_svm/linear_svm_function_impl.hpp [hunk 9]
```diff
@@ -259,11 +259,11 @@ void LinearSVMFunction<MatType, ParametersType>::Gradient(
   else
   {
     scores = parameters.rows(0, dataset.n_rows - 1).t() * dataset
-        + arma::repmat(parameters.row(dataset.n_rows).t(), 1,
+        + repmat(parameters.row(dataset.n_rows).t(), 1,
         dataset.n_cols);
```
Comment (rcurtin) [inline src/mlpack/methods/linear_svm/linear_svm_function_impl.hpp:263]:
```suggestion
        + repmat(parameters.row(dataset.n_rows).t(), 1, dataset.n_cols);
```

File: src/mlpack/methods/linear_svm/linear_svm_function_impl.hpp [hunk 10]
```diff
@@ -451,11 +451,11 @@ LinearSVMFunction<MatType, ParametersType>::EvaluateWithGradient(
   {
     scores = parameters.rows(0, dataset.n_rows - 1).t()
         * dataset.cols(firstId, lastId)
-        + arma::repmat(parameters.row(dataset.n_rows).t(), 1,
+        + repmat(parameters.row(dataset.n_rows).t(), 1,
                        (lastId - firstId + 1));
```
Comment (rcurtin) [inline src/mlpack/methods/linear_svm/linear_svm_function_impl.hpp:455]:
```suggestion
        + repmat(parameters.row(dataset.n_rows).t(), 1, (lastId - firstId + 1));
```

File: src/mlpack/methods/linear_svm/linear_svm_impl.hpp [hunk 11]
```diff
@@ -288,7 +288,7 @@ void LinearSVM<ModelMatType>::Classify(
   if (fitIntercept)
   {
     scores = parameters.rows(0, parameters.n_rows - 2).t() * data
-        + arma::repmat(parameters.row(parameters.n_rows - 1).t(), 1,
+        + repmat(parameters.row(parameters.n_rows - 1).t(), 1,
         data.n_cols);
```
Comment (rcurtin) [inline src/mlpack/methods/linear_svm/linear_svm_impl.hpp:292]:
```suggestion
        + repmat(parameters.row(parameters.n_rows - 1).t(), 1, data.n_cols);
```

File: src/mlpack/methods/linear_svm/linear_svm_impl.hpp [hunk 12]
```diff
@@ -314,7 +314,7 @@ void LinearSVM<ModelMatType>::Classify(
...
```
Comment (rcurtin) [inline src/mlpack/methods/linear_svm/linear_svm_impl.hpp:318]:
```suggestion
        + repmat(parameters.row(parameters.n_rows - 1).t(), 1, data.n_cols);
```

File: src/mlpack/methods/lsh/lsh_search_impl.hpp (hunk 5 above)
Comment (rcurtin) [inline src/mlpack/methods/lsh/lsh_search_impl.hpp:274]:
```suggestion
                                 this->referenceSet.n_cols);
```
No worries, we should at least align the second line though. :+1:

File: src/mlpack/methods/radical/radical_impl.hpp [hunk 13]
```diff
@@ -35,7 +35,7 @@ inline Radical::Radical(
 inline void Radical::CopyAndPerturb(arma::mat& xNew,
                                     const arma::mat& x) const
 {
-  xNew = arma::repmat(x, replicates, 1) + noiseStdDev * arma::randn(replicates * x.n_rows,
+  xNew = repmat(x, replicates, 1) + noiseStdDev * arma::randn(replicates * x.n_rows,
       x.n_cols);
```
Comment (rcurtin) [inline src/mlpack/methods/radical/radical_impl.hpp:39]:
```suggestion
  xNew = repmat(x, replicates, 1) + noiseStdDev * arma::randn(
      replicates * x.n_rows, x.n_cols);
```
This one was actually too long.

File: src/mlpack/methods/randomized_svd/randomized_svd_impl.hpp [hunk 14]
```diff
@@ -92,7 +92,7 @@ inline void RandomizedSVD::Apply(const MatType& data,
   if (data.n_cols >= data.n_rows)
   {
     R = arma::randn<arma::mat>(data.n_rows, iteratedPower);
-    Q = (data.t() * R) - arma::repmat(arma::trans(R.t() * rowMean),
+    Q = (data.t() * R) - repmat(arma::trans(R.t() * rowMean),
         data.n_cols, 1);
```
Comment (rcurtin) [inline src/mlpack/methods/randomized_svd/randomized_svd_impl.hpp:96]:
```suggestion
    Q = (data.t() * R) - repmat(arma::trans(R.t() * rowMean), data.n_cols, 1);
```

File: src/mlpack/methods/softmax_regression/softmax_regression_function_impl.hpp [hunk 15]
```diff
@@ -185,7 +185,7 @@ inline void SoftmaxRegressionFunction<MatType>::GetProbabilitiesMatrix(
         data.cols(start, start + batchSize - 1));
   }
 
-  probabilities = hypothesis / arma::repmat(arma::sum(hypothesis, 0),
+  probabilities = hypothesis / repmat(arma::sum(hypothesis, 0),
                                             numClasses, 1);
```
Comment (rcurtin) [inline src/mlpack/methods/softmax_regression/softmax_regression_function_impl.hpp:189]:
```suggestion
  probabilities = hypothesis / repmat(arma::sum(hypothesis, 0), numClasses, 1);
```

File: src/mlpack/methods/softmax_regression/softmax_regression_impl.hpp [hunk 16]
```diff
@@ -190,15 +190,15 @@ inline void SoftmaxRegression<MatType>::Classify(
...
       parameters.cols(1, parameters.n_cols - 1) * dataset);
   }
   else
   {
     hypothesis = arma::exp(parameters * dataset);
   }
 
-  probabilities = hypothesis / arma::repmat(arma::sum(hypothesis, 0),
+  probabilities = hypothesis / repmat(arma::sum(hypothesis, 0),
                                             numClasses, 1);
```
Comment (rcurtin) [inline src/mlpack/methods/softmax_regression/softmax_regression_impl.hpp:202]:
```suggestion
  probabilities = hypothesis / repmat(arma::sum(hypothesis, 0), numClasses, 1);
```

**Review by rcurtin — APPROVED** [review]
Looks good.  I found a bunch of places where we could still shorten lines (and one case where the line is too long).  But, those are all small style concerns.  Functionality-wise it looks great to me. :+1:

**Pull request merged**